f3/granite.go

```diff
@@ -350,8 +369,101 @@ func (i *instance) isValid(msg *GMessage) bool {
 	return true
 }
 
-// TODO Checks whether a message is justified by prior messages.
+// Checks whether a message is justified.
 func (i *instance) isJustified(msg *GMessage) bool {
+
+	if msg.Step == QUALITY_PHASE || msg.Step == PREPARE_PHASE {
+		//TODO check that the justification is Zero? necessary?
+		return true
+	}
+
+	if msg.Step == CONVERGE_PHASE {
+
+		//CONVERGE is justified by a strong quorum of COMMIT for bottom from the previous round.
+		// or a strong quorum of PREPARE for the same value from the previous round.
+		prevRound := msg.Round - 1
+		if msg.Justification.Round != prevRound {
+			i.log("dropping CONVERGE %v with evidence from wrong round %d", msg.Round, msg.Justification.Round)
+			return false
+		}
+
+		if msg.Justification.Step == PREPARE_PHASE.String() {
+			if msg.Value.HeadCIDOrZero() != msg.Justification.Value.HeadCIDOrZero() {
+				i.log("dropping CONVERGE for value %v with PREPARE evidence for a different value: %v", msg.Value, msg.Justification.Value)
+				return false
+			} else if msg.Value.IsZero() {
```

Comment:
using `if` instead of the `else if` when the previous statement returns would also make it easier to read

Reply:
Done.